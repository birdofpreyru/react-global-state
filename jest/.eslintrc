{
  "rules": {
    "import/no-extraneous-dependencies": ["error", {}],
    "no-console": "off"
  }
}
