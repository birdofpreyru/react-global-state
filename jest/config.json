{
  "collectCoverage": true,
  "collectCoverageFrom": ["src/**/*.{js,jsx,ts,tsx}"],
  "coverageDirectory": "__coverage__",
  "rootDir": "..",
  "testPathIgnorePatterns": [
    "/__assets__/",
    "/__tests__/ts-types/",
    "/node_modules/"
  ],
  "transform": {
    "\\.(j|t)sx?$": "babel-jest"
  }
}
