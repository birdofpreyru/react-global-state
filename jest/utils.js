/**
 * Jest test utils.
 */

import _ from 'lodash';
import mockdate from 'mockdate';
import { render, unmountComponentAtNode } from 'react-dom';

export { act } from 'react-dom/test-utils';

/**
 * Generates a mock UUID.
 * @param {Number} seed
 * @return {String}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => console.log.logs.push(_.cloneDeep(args));
  console.log.clear = () => {
    console.log.logs = [];
  };
  console.log.clear();
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {Number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return {HTMLElement}
 */
export function mount(scene) {
  const res = document.createElement('div');
  document.body.appendChild(res);
  render(scene, res);
  return res;
}

/**
 * Waits for `time`.
 * @param {Number} time Time [ms].
 * @returns {Promise} Resolves once `time` have passed.
 */
export async function timer(time) {
  return new Promise((resolve) => {
    setTimeout(resolve, time);
  });
}

/**
 * Unmounts `scene` from the DOM.
 * @param {HTMLElement} scene
 */
export function unmount(scene) {
  unmountComponentAtNode(scene);
  scene.remove();
}
