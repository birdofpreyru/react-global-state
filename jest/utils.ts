/**
 * Jest test utils.
 */

import { cloneDeep } from 'lodash';
import mockdate from 'mockdate';
import { type ReactNode } from 'react';
import { type Root, createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';

export { act };

(global as { [key: string]: unknown }).IS_REACT_ACT_ENVIRONMENT = true;

/**
 * Generates a mock UUID.
 * @param seed
 * @return
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

export const consoleLogs: unknown[] = [];

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => consoleLogs.push(cloneDeep(args));
  console.groupCollapsed = console.log;
  consoleLogs.splice(0);
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param time Time step [ms].
 * @returns Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

export type DestroyableHtmlElement = HTMLElement & { destroy: () => void };

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param scene
 * @return {HTMLElement} Created container DOM element with unmount
 *  function attached.
 */
export function mount(scene: ReactNode): DestroyableHtmlElement {
  let root: Root;

  const rootElement = document.createElement('div');

  const res: DestroyableHtmlElement = (rootElement as unknown) as DestroyableHtmlElement;

  res.destroy = () => {
    act(() => root.unmount());
    res.remove();
  };

  document.body.appendChild(rootElement);
  act(() => {
    root = createRoot(rootElement);
    root.render(scene);
  });
  return res;
}

/**
 * Advances mock timers by the specified time, and waits for any DOM updates
 * to complete.
 * @param time [miliseconds]
 * @return Wait for this.
 */
export function wait(time: number) {
  return act(() => mockTimer(time));
}
