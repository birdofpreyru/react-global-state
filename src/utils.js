// Auxiliary stuff.

/**
 * Returns 'true' if debug logging should be performed; 'false' otherwise.
 *
 * BEWARE: The actual safeguards for the debug logging still should read
 *  if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
 *    // Some debug logging
 *  }
 * to ensure that debug code is stripped out by Webpack in production mode.
 *
 * @returns {boolean}
 * @ignore
 */
export function isDebugMode() {
  try {
    return process.env.NODE_ENV !== 'production'
      && !!process.env.REACT_GLOBAL_STATE_DEBUG;
  } catch (error) {
    return false;
  }
}

export default null;
