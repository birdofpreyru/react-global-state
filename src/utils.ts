import { type GetFieldType } from 'lodash';

export type CallbackT = () => void;

// TODO: This probably should be moved to JS Utils lib.
export type TypeLock<
  Unlocked extends 0 | 1,
  LockedT extends never | void,
  UnlockedT,
> = Unlocked extends 0 ? LockedT : UnlockedT;

/**
 * Given the type of state, `StateT`, and the type of state path, `PathT`,
 * it evaluates the type of value at that path of the state, if it can be
 * evaluated from the path type (it is possible when `PathT` is a string
 * literal type, and `StateT` elements along this path have appropriate
 * types); otherwise it falls back to the specified `UnknownT` type,
 * which should be set either `never` (for input arguments), or `void`
 * (for return types) - `never` and `void` in those places forbid assignments,
 * and are not auto-inferred to more permissible types.
 *
 * BEWARE: When StateT is any the construct resolves to any for any string
 * paths.
 */
export type ValueAtPathT<
  StateT,
  PathT extends null | string | undefined,
  UnknownT extends never | undefined | void,
> = unknown extends StateT
  ? UnknownT
  : string extends PathT
    ? UnknownT
    : PathT extends null | undefined
      ? StateT
      : GetFieldType<StateT, PathT> extends undefined
        ? UnknownT : GetFieldType<StateT, PathT>;

export type ValueOrInitializerT<ValueT> = ValueT | (() => ValueT);

/**
 * Returns 'true' if debug logging should be performed; 'false' otherwise.
 *
 * BEWARE: The actual safeguards for the debug logging still should read
 *  if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
 *    // Some debug logging
 *  }
 * to ensure that debug code is stripped out by Webpack in production mode.
 *
 * @returns
 * @ignore
 */
export function isDebugMode(): boolean {
  try {
    return process.env.NODE_ENV !== 'production'
      && !!process.env.REACT_GLOBAL_STATE_DEBUG;
  } catch (error) {
    return false;
  }
}
