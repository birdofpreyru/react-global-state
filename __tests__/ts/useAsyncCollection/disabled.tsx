/** @jest-environment jsdom */

import type { FunctionComponent } from 'react';

import { mount } from 'jest/utils';

import {
  type AsyncDataEnvelopeT,
  newAsyncDataEnvelope,
  type UseAsyncDataResT,
  withGlobalStateType,
} from 'src';

type StateT = {
  collection: Record<string, AsyncDataEnvelopeT<string>>;
};

const {
  GlobalStateProvider,
  useAsyncCollection,
} = withGlobalStateType<StateT>();

const loader = jest.fn((id: string) => `res-${id}`);

let res: UseAsyncDataResT<string> | undefined;

function captureRes(value: UseAsyncDataResT<string>) {
  res = value;
}

const Component: FunctionComponent = () => {
  captureRes(useAsyncCollection('', 'collection', loader, { disabled: true }));
  return null;
};

const Scene: FunctionComponent = () => (
  <GlobalStateProvider
    initialState={{
      collection: { key: newAsyncDataEnvelope('value') },
    }}
  >
    <Component />
  </GlobalStateProvider>
);

test('useAsyncCollection(): "disabled" option works', () => {
  mount(<Scene />);
  expect(loader).not.toHaveBeenCalled();
  expect(res).toMatchSnapshot();
});
