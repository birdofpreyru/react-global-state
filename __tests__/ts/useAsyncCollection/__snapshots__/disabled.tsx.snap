// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`useAsyncCollection(): "disabled" option works 1`] = `
{
  "data": null,
  "loading": false,
  "reload": [Function],
  "set": [Function],
  "timestamp": 0,
}
`;
